cni/pkg/repair/netns.go

```diff
@@ -104,27 +96,31 @@ func getPodNetNs(pod *corev1.Pod) (string, error) {
 	}
 	oldest := uint64(math.MaxUint64)
 	best := ""
-	// We will iterate over all processes. Our goal is to find a process with the same network ID as we found above.
+
+	// We will iterate over all processes. Our goal is to find a process whose namespace has a veth with an IP matching the pod.
 	// There should be 1 or 2 processes that match: the pause container should always be there, and the istio-validation *might*.
 	// We want the pause container, as the istio-validation one may exit before we are done.
 	// We do this by detecting the longest running process. We could look at `cmdline`, but is likely more reliable to weird platforms.
 	for _, p := range procs {
 		ns := getPidNamespace(p.PID)
-		fd, err := unix.Open(ns, unix.O_RDONLY, 0)
+		id, err := vishnetns.GetFromPath(ns)
 		if err != nil {
-			// Not uncommon, many processes are transient and we have a TOCTOU here.
-			// No problem, must not be the one we are after.
-			log.Debugf("failed to open pid %v: %v", p.PID, err)
+			log.Warnf("failed to get netns for pid %v: %v", p.PID, err)
+			id.Close()
+			continue
+		}
+		defer id.Close()
+		if err := vishnetns.Set(id); err != nil {
```

Comment:
We can't just `Set` the netns - that will change the netns for the entire current thread, which is not safe.

probably need to use `netns.Do` here.

Reply:
Also `vishnetns.Set()` is deprecated in favor of  `golang.org/x/sys/unix.Setns`

Reply:
Ack.  So to summarize the changes you're suggesting:
1. At the start of the function, call runtime.LockOSThread() and save the original netns 
2. At the end of the function, restore original netns and call runtime.UnlockOSThread().

Since we're using runInHost / netns.Do, won't Lock/Unlock be called implicitly here?

Also can you point to the ambient logic to compare?

Reply:
- `runInHost / netns.Do` -> locks us to this thread, moves this thread to host netns 
- restores original netns and does a thread unlock after the `Do` closure completes, signaling we are safe to carry on.

`netns.Do` has this caveat: 

>      // Executes the passed closure in this object's network namespace,
> 	// attempting to restore the original namespace before returning.
> 	// However, since each OS thread can have a different network namespace,
> 	// and Go's thread scheduling is highly variable, callers cannot
> 	// guarantee any specific namespace is set unless operations that
> 	// require that namespace are wrapped with Do().  Also, no code called
> 	// from Do() should call runtime.UnlockOSThread(), or the risk
>	// of executing code in an incorrect namespace will be greater.  See
>	// https://github.com/golang/go/wiki/LockOSThread for further details.

It should be fine to nest `netns.Do`s (one for hostns, one for each pod ns) since that will make successive `runtime.LockOSThread()` + `runtime.UnlockOSThread()` calls, and Go will trash the thread if you don't make exactly as many Unlocks as you do Locks.

Reply:
> Also can you point to the ambient logic to compare?

It mostly does the same thing but using the same helpers in both spots is probably a bit out of scope for this, so don't worry too much about that, that's just me remarking to myself.

In ambient we are either given the netns by the CNI on pod start, or (if the pod is already running, as it would be with sidecar repair) we scan `/proc`: `istio/cni/pkg/nodeagent/podcgroupns.go`